lib/schema/duplicate.js

```diff
@@ -0,0 +1,77 @@
+'use strict';
+
+var amoeba = require('amoeba');
+var except = amoeba.except;
+var misc = require('../misc.js');
+var schema = require('./schema.js');
+
+var idHashFieldMap = {};
+
+const getIdHashFields = function (type) {
+  var retVal = idHashFieldMap[type];
+  if (retVal == null) {
+    throw except.IAE('No known hashFields for type[%s]', type);
+  }
+  return retVal;
+};
+
+exports.registerFieldsForDuplicator = function (type, idHashFields = []) {
+  if (idHashFieldMap[type] == null) {
+    idHashFieldMap[type] = [
+      '_userId',
+      'deviceId',
+      'time',
+      'type',
+      ...idHashFields,
+    ];
+  } else {
+    throw except.IAE(
+      'Id hash fields for type[%s] already defined[%j], cannot set[%j]',
+      type,
+      idHashFieldMap[type],
+      idHashFields
+    );
+  }
+};
+
+exports.generateHash = function (datum) {
+  if (typeof datum === 'string') {
+    return datum;
+  }
+  var idHashFields = idHashFieldMap[datum.type];
+  if (!idHashFields) {
+    return '';
```

Comment:
Does the caller check for empty string and not write the platform deduplicator hash? We don't want any with empty strings in the database I think.